### ISSUE TITLE
Add support for termux

### ISSUE THREAD

**Issue by JesusChapman** [hwittenborn/makedeb#61]
### Description of feature

Termux is a terminal for android that does not require root, it uses apt and dpkg to manage its native packages for android.

More info :

- https://wiki.termux.com/wiki/Differences_from_Linux

- https://github.com/termux/termux-packages/wiki/Termux-file-system-layout

The only thing you have to do, to make it work in termux, is change the PREFIX of makedeb from (/usr) to (/data/data/com.termux/files/usr)

![Screenshot_20211126-153239](https://user-images.githubusercontent.com/64093255/143627554-d8a16450-f65b-4012-94bb-3cd009557b86.png)

**Note:**
Pacman is available for use by makepkg, but there are no packages, all available packages are .deb and use the $PREFIX of termux

### Installed makedeb version

No installed

**Comment by hwittenborn** [comment]
Just two question:

- Is this going to require packaging changes (i.e. I need to modify the packaged archives to install files at `/data/data/com.termux/files/usr` instead of `/usr`), or would Termux handle that all itself?

- I'm assuming I need to change any hardcoded paths in makedeb to be relative or to just point to the paths Termux uses, or would that also be automatic?

**Comment by JesusChapman** [comment]
> 

yes, you must change the makedeb paths to use the termux prefix, if you use the standard paths (/usr) it won't work.

**Comment by JesusChapman** [comment]
Example 

```bash
####################
##  BEGIN SCRIPT  ##
####################
# Get makepkg message syntax
source "/data/data/com.termux/files/usr/share/${makepkg_package_name}/util/message.sh"
colorize
```

**Comment by JesusChapman** [comment]
note that if your program uses /proc stuff, or /dev if it should use the standard path and not the termux prefix, since it is android.

but those things will only have read access

the links I left in the first comment contain a lot of information on this.

**Comment by hwittenborn** [comment]
If I have to create a separate archive or anything just for Termux packaging, that would probably need to be maintained by a fork. I don't use Termux myself, so I can't really support it either, nor do I see myself much at all wanting to.

Would there be any other methods to get this done? If it would have to involve a separate package for Termux, I could work on getting the Makefile in a shape to specify directories for compatibility with Termux though, but I don't see myself supporting this in the official APT repository if it requires such.

**Comment by JesusChapman** [comment]
> If I have to create a separate archive or anything just for Termux packaging, that would probably need to be maintained by a fork. I don't use Termux myself, so I can't really support it either, nor do I see myself much at all wanting to.
> 
> Would there be any other methods to get this done? If it would have to involve a separate package for Termux, I could work on getting the Makefile in a shape to specify directories for compatibility with Termux though, but I don't see myself supporting this in the official APT repository if it requires such.

If makedeb runs smoothly, I can add the deb packages to the repositories.

termux has a Docker environment and a cross-compilation system

**Comment by JesusChapman** [comment]
Like many mprweb packages should work without problems, since the compilation steps in termux are the same as in Linux, the only thing that changes is the prefix.

and since the PKGBUILD use a variable when installing "package ()", if makedeb has the variables pointing to the termux prefix, nothing would have to be rewritten, the packages would work anyway without any problem.

**Comment by hwittenborn** [comment]
Just double checking the Makefile and it looks like it'd be ready to be used in custom environments, the only problem being I need to get the Makefile into the `stable` branch first (I'm assuming you'd like to use the stable release of makedeb for packaging?).

If you'd like to experiment with that off of the `alpha` branch for the time being it should just be a drop-in replacement if you ever go to the `stable` branch later.

**Comment by JesusChapman** [comment]
> Simplemente verificando dos veces el Makefile y parece que estaría listo para usarse en entornos personalizados, el único problema es que `stable`primero necesito obtener el Makefile en la rama (supongo que le gustaría usar la versión estable de makedeb para el embalaje?).
> 
> Si desea experimentar con eso fuera de la `alpha`sucursal por el momento, debería ser un reemplazo directo si alguna vez va a la `stable`sucursal más tarde.

better use stable branch

**Comment by hwittenborn** [comment]
> Example
> 
> ```shell
> ####################
> ##  BEGIN SCRIPT  ##
> ####################
> # Get makepkg message syntax
> source "/data/data/com.termux/files/usr/share/${makepkg_package_name}/util/message.sh"
> colorize
> ```

Looks like I completely glanced over this message earlier :P

That'll indeed require some changes in makedeb to support such, but it's nothing that wouldn't be more than an option for the Makefile. I'll let you know once I add support to make that all work.

**Comment by JesusChapman** [comment]
Thanks

**Comment by hwittenborn** [comment]
Is this still something you're interested in getting done @Yisus7u7? I'd like to start a PR for this, but I'd probably need you to verify that everything works on your end before merging it.

**Comment by JesusChapman** [comment]
> Is this still something you're interested in getting done @Yisus7u7? I'd like to start a PR for this, but I'd probably need you to verify that everything works on your end before merging it.

What do you need?

**Comment by Aeres-u99** [comment]
@Yisus7u7 could you verify a package or two which are source based installable? (-git packages) let's see if those would work in termux environment. Certainly, @hwittenborn we can have a workaround where in instead of hardcoding the variables to /usr we can take and deduce it from environment (like identify if its termux) this might come in handy as a chroot rescue method as well, wherein we could modify the installation path. For a crude example, let's say I broke my debian and I need to fix it, then all I would have to do is from usb environment mount the system, and set the installation path to the mounted location which would basically allow us to install package at the location instead of the system.

**Comment by Aeres-u99** [comment]
Also, @Yisus7u7 forgive me if it has been discussed, but have you tried running mpr inside a proot container on termux? that should possibly work ootb without needing any modifications in the code.

**Comment by JesusChapman** [comment]
> @Yisus7u7 could you verify a package or two which are source based installable? (-git packages) let's see if those would work in termux environment. Certainly, @hwittenborn we can have a workaround where in instead of hardcoding the variables to /usr we can take and deduce it from environment (like identify if its termux) this might come in handy as a chroot rescue method as well, wherein we could modify the installation path. For a crude example, let's say I broke my debian and I need to fix it, then all I would have to do is from usb environment mount the system, and set the installation path to the mounted location which would basically allow us to install package at the location instead of the system.

Yes, termux has git and many other necessary packages.

**Comment by JesusChapman** [comment]
> Also, @Yisus7u7 forgive me if it has been discussed, but have you tried running mpr inside a proot container on termux? that should possibly work ootb without needing any modifications in the code.

I haven't tried it, but I still don't like to use proot because of its performance, termux already offers native packages.

**Comment by hwittenborn** [comment]
I still need to fix a few things up, I added an implementation, but it's kind of messy now that I'm looking at it. I'll let you know when something's good for it.

**Comment by JesusChapman** [comment]
> I still need to fix a few things up, I added an implementation, but it's kind of messy now that I'm looking at it. I'll let you know when something's good for it.

Thanks!

**Comment by hwittenborn** [comment]
Nevermind, it looks like the current implementation was what I was going after.

So when using the Makefile to build, you need to pass `LIBRARY_DIR` so makedeb knows where to look for the library files. From what I'm seeing it should probably be `LIBRARY_DIR='/data/data/com.termux/files/usr/share/makedeb`.

Unless Termius doesn't accept binary packages, I'd just build from the MPR with the `LIBRARY_DIR` argument added to the `make` call added in the `prepare()` function. Regardless of if you do that you'll need it added to your `make prepare` call though.

Let me know if anything doesn't work right.

**Comment by JesusChapman** [comment]
@hwittenborn 

![Screenshot_20220404-162901](https://user-images.githubusercontent.com/64093255/161635379-43a6e79b-cde3-41a8-acde-f0820a637dbe.png)

**Comment by hwittenborn** [comment]
Oh, I didn't add those parts :P

I'm wrapping up a PR rn, then I'll be right on to that

**Comment by hwittenborn** [comment]
This should be fixed in #149 @Yisus7u7 - would you be able to try installing from that branch and see if everything works fine for you in Termux?

**Comment by JesusChapman** [comment]
![Screenshot_20220404-213906](https://user-images.githubusercontent.com/64093255/161668247-6809b31b-784c-437b-80d2-2c03c204c4f7.png)

**Comment by JesusChapman** [comment]
It still relies on standard Linux paths to work.

![Screenshot_20220404-213959](https://user-images.githubusercontent.com/64093255/161668382-d4eea199-79b6-4782-b3eb-99531dd110b6.png)

**Comment by hwittenborn** [comment]
Oh, I would've sworn I sent over the new command need in the Makefile, it looks like I didn't;

You should need to build with `FILESYSTEM_PREFIX` instead of the previous `LIBRARY_DIR`. That's what I'm using to make all of makedeb's files go into a new central prefix.

Sorry about that :p

**Comment by JesusChapman** [comment]
Well, I'll be watching

**Comment by hwittenborn** [comment]
Sorry, I meant you need to change that on your end. Everything *should* be implemented correctly in the Makefile currently, you should just need to swap that variable name during your build

**Comment by JesusChapman** [comment]
Thanks.

**Comment by hwittenborn** [comment]
Is it working for you now @Yisus7u7?

**Comment by hwittenborn** [comment]
Going to go ahead and close this since there hasn't been a response. This got merged in #149, and should be in `stable` soon.

Send a message if it doesn't work still @Yisus7u7.

**Comment by JesusChapman** [comment]
> Going to go ahead and close this since there hasn't been a response. This got merged in #149, and should be in `stable` soon.
> 
> Send a message if it doesn't work still @Yisus7u7.

Ok, I came back, I was busy for a while.
